File: app/src/main/java/com/songlink/share/api/SonglinkClient.java
package com.songlink.share.api;

import com.songlink.share.BuildConfig;
import com.songlink.share.model.SonglinkResponse;
import android.util.Log;
import org.json.JSONObject;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;

public class SonglinkClient {

    public static class ApiException extends Exception {
        private final int code;

        public ApiException(int code) {
            super("HTTP " + code);
            this.code = code;
        }

        public int getCode() {
            return code;
        }
    }

    private static final String ENDPOINT = "https://api.song.link/v1-alpha.1/links";

    public static SonglinkResponse getLinks(String url) throws Exception {
        String encoded = URLEncoder.encode(url, "UTF-8");
        HttpURLConnection connection = (HttpURLConnection) new URL(ENDPOINT + "?url=" + encoded).openConnection();
        try {
            connection.setConnectTimeout(15000);
            connection.setReadTimeout(30000);

            int code = connection.getResponseCode();
            if (BuildConfig.DEBUG) {
                Log.d("SonglinkClient", "→ " + code + " for " + url);
            }

            if (code != HttpURLConnection.HTTP_OK) {
                throw new ApiException(code);
            }

            BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), "UTF-8"));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }
            reader.close();

            JSONObject json = new JSONObject(sb.toString());
            String pageUrl = json.optString("pageUrl", null);
            if (pageUrl != null && pageUrl.isEmpty()) {
                pageUrl = null;
            }
            return new SonglinkResponse(pageUrl);
        } finally {
            connection.disconnect();
        }
    }
}

File: app/src/main/kotlin/com/songlink/share/api/SonglinkClient.kt
package com.songlink.share.api

import android.util.Log
import com.songlink.share.BuildConfig
import com.songlink.share.model.SonglinkResponse
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

class ApiException(val code: Int) : Exception("HTTP $code")

object SonglinkClient {

    private const val ENDPOINT = "https://api.song.link/v1-alpha.1/links"

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        coerceInputValues = true
    }

    suspend fun getLinks(url: String): SonglinkResponse = withContext(Dispatchers.IO) {
        val encoded = URLEncoder.encode(url, "UTF-8")
        val connection = URL("$ENDPOINT?url=$encoded").openConnection() as HttpURLConnection
        try {
            connection.connectTimeout = 15_000
            connection.readTimeout = 30_000

            val code = connection.responseCode
            if (BuildConfig.DEBUG) Log.d("SonglinkClient", "→ $code for $url")

            if (code != HttpURLConnection.HTTP_OK) throw ApiException(code)

            val body = connection.inputStream.bufferedReader().readText()
            json.decodeFromString<SonglinkResponse>(body)
        } finally {
            connection.disconnect()
        }
    }
}
